[WIP] 2.x Raw meta

gchtr opened the pull request:
**Ticket**: #1823

This pull request is work in progress and shouldn’t be merged yet.

## Issue

See #1823.

## Solution

This pull request makes changes to the `Post`, `Term`, `User`, `Comment` and `MenuItem` classes and updates the meta logic to work in the same way.

### New `raw_meta()` method

There’s a new method `raw_meta()`, for accessing unfiltered, raw meta values from the database.

### The `$custom` property

I changed the visibility of the `$custom` property to `protected`, because it shouldn’t be accessed directly. To access raw meta values, `raw_meta()` should be used.

I added missing `$custom` properties for the `Term`, `User` and `Comment` classes.

### Do not import meta values as dynamic properties

Meta values shouldn’t be imported to the object as dynamic properties, so I made sure that the meta values are not imported using `Core::import()`. Instead, as suggested in https://github.com/timber/timber/issues/478#issuecomment-443916908, when we see something like `{{ post.my_field }}`, then that

> … tells me that the theme author HAS written some kind of special processing for this field and I should refer to the subclass for appropriate info.

### Caching

In https://github.com/timber/timber/issues/1823#issuecomment-438447332, I wrote:

> Currently, we don’t check for an already retrieved value in `Timber\Post::meta()`. We always get it from the database through `get_post_meta()`. Do we really get raw values for free? What if somebody only wants to access one meta value of a post that should be filtered in the template? Will the population of the object with raw values run for nothing?
> 
> I see two ways to go here:
> 
> 1. Deprecate that the post object gets populated with raw meta values on initialization and only use `raw_meta()` to access raw properties. Let WordPress do all the performance optimization.
> 2. Keep fetching all raw meta values, and optimize `meta()` to access the raw values instead of fetching them from the database. I guess there would be no drawbacks, because we’d use `get_post_custom()`, which directly uses `get_post_meta()`.

I checked how WordPress does fetch meta values. It will always fetch all meta values for a post and save it in the object cache the first time any meta value is requested. This means that when we fetch all custom meta data on post initialization trough `get_post_meta()`, there’s **no performance hit** when there are further requests to meta values, e.g. through `$post->meta( 'field_name' )`.

This means that my 2 options above are both invalid. We don’t need to «optimize `meta()` to access the raw values instead of fetching them from the database», because it’s already optimized by WordPress itself.

If developers wanted to disable fetching the meta values on object initialization by default, they could still do that via the `timber/post/pre_get_meta_values` filter (see #1903).

```php
add_filter( 'timber/post/pre_get_meta_values', '__return_false' );
```

Or they could provide their own method for fetching the values. In that case, they’d have to think about caching themselves.

### Documentation

I updated the documentation with the new approach and also added a new section in the Upgrade Guide for 2.0.

## Impact

- Errors will be thrown when the `$custom` property is accessed directly.
- There should be no impact for accessing dynamic properties (`{{ post.my_custom_field }}`) based on what is discussed in #478.

## Usage Changes

- Use `{{ post.raw_meta('my_custom_field') }}` instead of `{{ post.custom.my_custom_field }}`.
- Use `{{ post.meta('my_custom_field') }}` instead of `{{ post.my_custom_field }}`.

## Considerations

### Change the name of `$custom` to `$meta_values`

The property for meta values is called `$custom`. Everything is else related to meta values in the classes is called something including "meta". The word "custom" refers to custom fields, which might save values in the meta database. But since the `$custom` holds raw values from the meta database, shouldn’t the property be called `$meta_values`? I think this would be more consistent.

### Error message when trying to access `$custom` property

Currently, there’s no dedicated error message when trying to access the protected `$custom` property. Maybe we can catch this in the `__get()` call to help developers use the new preferred way to access meta values through `meta()` and `raw_meta()`.

### How to check for an existing post ID

Currently, the methods to find the proper post IDs are not very consistent throughout the different classes. That’s why I added a check for existing IDs with a bailout to all the `get_meta_values()` methods. For example, in `Post::get_meta_values()`, I had to add

```php
if ( ! $pid ) {
    return null;
}
```

Possibly, this can be improved with the introduction of object factories, where an object can only be instantiated with a proper ID.

## Testing

Before I can fix existing tests and add more tests for the new `raw_method()`, we should probably resolve #478 first. There’s a lot of tests that don’t fail, although they try to access the now protected `$custom` property, because of the following check in `Core::__get()`:

```php
if ( property_exists( $this, $field ) ) {
    return $this->field;
}
```

## Todo

- [x] Resolve #478.
- [x] Add tests.

gchtr commented:
@acobster @pascalknecht Your review was requested automatically because I added more commits. Please don’t review yet, it’s still work in progress. I still need to add more tests.

gchtr commented:
Okay, so before I can finish this up, I have some thoughts that I want to discuss. But first, a summary of what changed until here.

## Changes to the original pull request

### Error message when trying to access `$custom` property

> Currently, there’s no dedicated error message when trying to access the protected `$custom` property. Maybe we can catch this in the `__get()` call to help developers use the new preferred way to access meta values through `meta()` and `raw_meta()`.

I added a new check in `Core::__get()` that will throw a warning when somebody tries to access a value through the `$custom` property.

### Optional parameter for `meta()`

The first parameter `$field_name` in `Post::meta()` was optional, but not in all other objects, so I removed it in `Post::meta()`. Let me get back to this decision later.

### Imported meta values

I found two more cases in `Timber\User` and `Timber\Comment` where meta values were imported into the object and I removed those imports. This will streamline the meta behavior accross the different classes that use meta.

### Removed ACF filter when fetching term fields

There was a filter `timber/term/get_meta_values` in the ACF integration that fetched meta values through ACF. But this filter is used in the function that is supposed to fetch raw values, so ACF shouldn’t act here. That’s why I removed the filter.

### Updated tests for different use cases related to access through magic methods

There should be no impact for accessing dynamic properties (`{{ post.my_custom_field }}`) based on what is discussed in #478. However, there are a couple of cases that are special:

1. Trying to access a custom field name that is also the name of a **public method** on the object. In this case, the result of the method should take precedence over the value of the custom field.
2. Trying to access a custom field name that is also the name of a **protected method** on the object. In this case, the value of the custom field should take precedence over the value of the method.
2. Trying to access a custom field name that is also the name of a **public method on the object that has required parameters**. In this case, there will be an `ArgumentCountError`.
3. Trying to access a custom field name that is also the name of a **public property** on the object. In this case, the value of the property should take precedence over the value of the custom field.
4. Trying to access a custom field name that is also the name of a **protected property** on the object. In this case, the value of the custom field should take precedence over the value of the property.

I added a couple of new tests for the different scenarios listed above. Because most of the tests are very similar, I added all the tests in a single `TestTimberMeta` class. To make it easier to check the differences between public and protected methods and properties, I added helper classes: `MetaPost`, `MetaTerm`, `MetaUser` and `MetaComment`. I hope that like this, the tests will be easier to maintain.

### Documentation

I updated the documentation again to make it clearer how fetching custom values works.

## Discussion

When I worked on the tests for this pull request, I realized that I missed two big points that also need be considered:

### 1. Disabling pre-fetch

When it’s possible to disable the pre-fetching of meta values through the `timber/{$object}/pre_get_meta_values` filter, then `raw_meta()` should still return a value. Currently, it doesn’t. So we need to check whether meta values already exist on an object in `raw_meta()`, and if it doesn’t, we need to load them.

I then asked myself: **Why do we ever have to load meta values on object initialization?** If we drop that and lazy-load meta values only if they are needed, we easily get better performance. There would be no need for developers to disable pre-fetching through a filter. That’s probably heading into the direction that @aj-adl already mentioned.

### 2. Getting all meta values through raw_meta()

Jared already mentioned it in in the first point in #1823, but I somehow missed it: It should be possible to load all meta values through `raw_meta()` by not providing a parameter. This way, the method would replace `{{ dump(post.custom) }}`. We could instead write `{{ dump(post.raw_meta) }}` to get a list of all meta values set on an object.

The same probably goes for `meta()`. Instead of making `$field_name` a required parameter in all objects – like they are now – we should probably make them optional for all objects.

I then asked myself: What’s the reason for `$post->custom` to exist at all? It’s clear that it would act as a small cache for meta values fetched on a post. But then again, fetching meta data is already properly cached by WordPress and we probably don’t need to over-optimize here. We don’t do it for `$post->meta()` either, why would we have to do it for `$post->raw_meta()`? The only thing that might hit performance are the filters that are applied.

So what if we removed the `$custom` property and renamed `$post->get_meta_values()` to `$post->raw_meta()`? Deprecation notices for the `$custom` property are already there in `Core::_get()`.

## Plan

Based on the points listed above, I want to make the following changes to this pull request.

### meta()

- Make it optional to pass a `$field_name` parameter for `meta()`. Not providing a field name would mean that Timber fetches **all meta values**.

### raw_meta()

- Remove pre-fetching of meta values from `init()` functions.
- Remove `$custom` property.
- Rename `get_meta_values()` function to `raw_meta()`. Don’t keep an internal cache for meta values, but let WordPress do the caching of meta values.
- Make it optional to pass a `$field_name` parameter for `raw_meta()`. Not providing a field name would mean that Timber fetches **all meta values**, with values not filtered by third party plugins like ACF.

---

What do you all think of this direction? @jarednova @acobster @pascalknecht @aj-adl @palmiak.

jarednova commented:
@gchtr I think this is all well-reasoned and feels like the cleanest solution to the variety of issues we've seen here. The one thing I've tried before with some issue that you should be aware of:

```
meta()
Make it optional to pass a $field_name parameter for meta(). Not providing a field name would mean that Timber fetches all meta values.
```

... I remember testing this back in the day and found it was a major performance hit (because when ACF or another filter processes those all at once, there can be a ton of attachment objects that also get queried). I can see how it's valuable for development or debugging — but we might want to acknowledge as much in the documentation (it's also possible that things have gotten better since I ran this test a few years ago).

Similarly, I agree with the handling of the "special cases" you outline — glad the tests are in there to preserve that logic so we don't have to stress about it again in the future

gchtr commented:
@jarednova Yes, we would definitely have to make it clear in the documentation that it can have a huge performance hit if all values are fetched at once.

I looked at it a little closer and I think we wouldn’t even need a full-fledged `raw_meta()` function. The two functions would practically do the same, except `meta()` would apply an additional `timber/{$object}/meta` filter. What if `raw_meta()` looked like this?

```php
public function raw_meta( $field_name = '', $args = [] ) {
    return $this->meta( $field_name, array_merge(
        $args,
        [ 'raw' => true ]
    ) );
}
```

jarednova commented:
@gchtr — haven't run it, but that looks right to me! My only Q is whether in finalizing this, we should make "rawness" true or false, that is ...

### Raw is `true`
```php
public function raw_meta( $field_name = '', $args = [] ) {
    return $this->meta( $field_name, array_merge(
        $args,
        [ 'raw' => true ]
    ) );
}
```

Depending on what goes on inside `meta()` to return the "raw" version, is it more accurate to say:

### Filter/Process is `false`
```php
public function raw_meta( $field_name = '', $args = [] ) {
    return $this->meta( $field_name, array_merge(
        $args,
        [ 'apply_filters' => false ]
    ) );
}
```

I know it's a bit pedantic, but just want to best represent what's going on internally

gchtr commented:
@jarednova Good input! I like `apply_filters` better, too.

I first thought that we might only skip the `timber/post/meta` filter and still apply the `timber/post/pre_meta` filter for `raw_meta()`. But if I add a `timber/post/pre_meta` filter with some custom filtering and use `raw_meta(), I should still expect a value directly from the database and never a filtered value.

Cool, I have everything I need to keep working on this 👍.

gchtr commented:
Finally. Here are the changes since my last comment:

### Remove $custom property and automatic import of meta values

I removed the `$custom` property for all objects and the automatic import of post meta values on object initialization. Because of this, I could remove some unnecessary tests.

### Remove `get_meta_values()` method 

Because there’s no automatic import of meta values anymore, there’s no need for the `get_meta_values()` function and some of the filters that cam with it. I removed the `get_meta_values()` method for all objects and merged the logic with the `meta()` function.

The following filters that were introduced earlier in 2.x were removed again:

- `timber_post_get_meta_pre`, use `timber/post/pre_get_meta_values` instead
- `timber_post_get_meta`, use `timber/post/get_meta_values` instead
- `timber_term_get_meta`, use `timber/term/get_meta_values` instead
- `timber_comment_get_meta_pre`, use `timber/comment/pre_get_meta_values` instead
- `timber_comment_get_meta`, use `timber/comment/get_meta_values` instead
- `timber_user_get_meta_pre`, use `timber/user/pre_get_meta_values` instead
- `timber_user_get_meta`, use `timber/user/get_meta_values` instead

Remember how we discussed [better handling in `timber/post/pre_get_meta_values`](https://github.com/timber/timber/issues/2011)? It’s now `timber/post/pre_meta` only. It would be great, if I could get some feedback of you, @aj-adl.

### Update order in `timber/user/meta` filter

In `timber/user/meta`, the last two arguments were switched. The user object comes before the arguments to match the order of the meta filter in the other classes.

### Deprecate Post::$_thumbnail_id property

The `Post::$_thumbnail_id` property is a property that was populated when meta values were imported into the object. Because that functionality is deprecated, we need to deprecate that property as well.

In order to fix failing tests for functions that relied on meta values related to this property, I figured it would be good to add a `Post::thumbnail_id()` method as a clear way to get the thumbnail ID.

To have a deprecation notice, I added a check to `Post::__get()` throws a deprecation notice, including tests.

### Deprecate MenuItem::$_menu_item_object_id

Similar to the `Post::$_thumbnail_id`, all `_menu_item…` properties still work through magic methods. In MenuItem::master_object(), the `$custom` property was still used. I added a new public property `MenuItem::$object_id`, including a test.

### Make filters in ACF Integration removable

I made the method calls in the ACF integration static so that the filters can be removed for testing. It’s important that ACF filters are not run when testing the meta filters, because they would interfere with the logic.

## Considerations

- Should we add methods or properties for all meta values that are saved for menu items and for posts (like post thumbnails) by default?
- Do we need to add `convert()` to all objects? Currently it’s only used in `Post::meta()`. But that logic might also be related to the Object Factories, so we shold probably wait with this.

## Tests

I worked a lot on the tests and restructured them some more.

@jarednova Now you can a close look ;)

jarednova commented:
I think we're looking good @gchtr! I made a few spelling/grammar fixes in my commit above. But overall, I think I just have two questions ...

## Revisions/Previewing
It looks like a chunk of code from `Post` relating to previewing a post revision got moved into `Post:init()`, however I can't tell where/if the results of this code are ever used. It would seem that the test: `TestTimberRevisions::testCustomFieldPreviewRevisionMethod` covers this code. The good news is that this test passes! The (maybe) bad news is that when commenting out this part, the test also passes. Is this just a dead section we can delete?

## Default Properties/Methods
You asked:
> Should we add methods or properties for all meta values that are saved for menu items and for posts (like post thumbnails) by default?
My thought here is "no," or at least "not yet" — while it makes some intuitive sense that these could be present, I think aspiring for "completeness" introduces extra code/documentation that doesn't prove many benefits. Inevitably, it becomes another thing to manage during merges and upgrades which can reduce our throughput. I'd prefer an approach of "incorporation" — where we update on an as-needed basis.

## Convert
You asked:
> Do we need to add convert() to all objects? Currently it’s only used in Post::meta(). But that logic might also be related to the Object Factories, so we shold probably wait with this.
Agreed. It seems to me this is handled by the object factories approach; or at least we should wait until object factories is settled

jarednova commented:
@gchtr just a few things here for you to address. @pascalknecht @acobster do either of you want to take a look too?

gchtr commented on lib/Post.php:
```diff
@@ -401,6 +400,13 @@ protected function init( $pid = null ) {
+		if ( $this->is_previewing() ) {
+			global $wp_query;
+			$post_custom_id = $this->get_post_preview_id( $wp_query );
+		}
```
I don’t know anymore, either. This part was probably used to import custom meta data. I only removed the function call to import the meta data, but not the logic that figures out the post ID to use to fetch the meta data.

I went ahead and removed that part.

jarednova replied:
Cool — I ran a test and "previewed" content all seemed to work as intended even w/o this section

jarednova commented:
@pascalknecht @acobster — @gchtr was able to take care of the issue I flagged. Let me know if you'd like to give this a look as well, otherwise I'll pull in to the `2.x` branch in the next few days
